simulator/bypass/data_bypass.h

```diff
@@ -9,172 +9,251 @@
-                auto get_register_num() const { return register_num; }    
+                auto get_register_num() const { return register_num; }
 
             private:
                 const RegisterStage bypassing_stage;
-                const MIPSRegister register_num;
+                const Register register_num;
         };
 
+        // checks whether the source register of passed instruction is in RF  
+        auto is_in_RF( const FuncInstr& instr, uint8 src_index) const
+        {
+            const auto reg_num = instr.get_src_num( src_index);
+            return scoreboard.get_entry( reg_num).current_stage == RegisterStage::IN_RF;
+        }
+
+        // checks whether the source register of passed instruction is bypassible
+        auto is_bypassible( const FuncInstr& instr, uint8 src_index) const
+        {
+            const auto reg_num = instr.get_src_num( src_index);
+            return scoreboard.get_entry( reg_num).is_bypassible;
+        }
+
+        // checks if the stall needed for passed instruction
+        auto is_stall( const FuncInstr& instr) const
+        {
+            return ( (!is_in_RF( instr, 0) && !is_bypassible( instr, 0)) ||
+                     (!is_in_RF( instr, 1) && !is_bypassible( instr, 1)) );
+        }
+
+        // returns bypass command for passed instruction and its source register
+        // in accordance with current state of the scoreboard
+        auto get_bypass_command( const FuncInstr& instr, uint8 src_index) const
+        {
+            const auto reg_num = instr.get_src_num( src_index);
+            return BypassCommand( get_current_stage( reg_num), reg_num);
+        }
+
+        // returns an index of the port where bypassed data should be get from
+        // in accordance with passed bypass command
+        auto get_bypass_direction( const BypassCommand& bypass_command) const
+        {
+            const auto bypassing_stage = bypass_command.get_bypassing_stage();
+            return static_cast<uint8>( bypassing_stage);
+        }
+
+        // transforms bypassed data if needed in accordance with passed bypass command
+        static uint64 adapt_bypassed_data( const BypassCommand& bypass_command, uint64 bypassed_data);
+
+        // introduces new instruction to bypassing unit
+        void trace_new_instr( const FuncInstr& instr);
+
+        // updates the scoreboard
+        void update();
+
+        // removes the information about passed instruction from the scoreboard
+        void untrace_instr( const FuncInstr& instr);
+    
     private:
         struct RegisterInfo
         {
-            RegisterStage current_stage = RegisterStage::in_RF();
-            RegisterStage ready_stage = RegisterStage::in_RF();
+            RegisterStage current_stage = RegisterStage::IN_RF;
+            RegisterStage ready_stage = RegisterStage::IN_RF;
             bool is_bypassible = false;
+            bool is_traced = false;
         };
 
 
         class Scoreboard
```

Comment:
Do we really need `Scoreboard` class? It has no other functionality than `std::array` has.